### PR TITLE
Fix shopping cart when allow checkout with wrongly product

### PR DESCRIPTION
**Pull request by kompilorb**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | The shopping cart allows checkout with inactive product or product whose quantity is upper than available stock.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-3268
| How to test?  | Add a product to cart and don't checkout, go to BO > Catalog > Products and set the product inactive or set its quantity to 0. Then go to shopping cart, you will see the product is still in the cart and when you click on "PROCEED TO CHECKOUT", check if an error message appears.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8478)

### CODE REVIEW

**Comment by LittleBigDev** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[controllers/front/CartController.php, line 85 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPrW8EybfdUKw7Wcrs:-KyPrW8EybfdUKw7Wcrt:b-z7mum2) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L85)):*
> ```PHP
>             } else if (Tools::getIsset('checkout')) {
>                 Tools::redirect($this->context->link->getPageLink('order'));
>             }
> ```

I would change the order, and get rid of the `elseif` :
```php
if (Tools::getIsset('checkout')) {
    return Tools::redirect($this->context->link->getPageLink('order'));
}

if (true !== $isAvailable) { // Yoda condition
    $this->errors[] = $isAvailable;
}
```

---

*[controllers/front/CartController.php, line 136 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPunTspzXxSgqYwM4e:-KyPunTspzXxSgqYwM4f:bj4k4n5) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L136)):*
> ```PHP
>                 'quantity' => $productQuantity,
>                 'cart' => $cartPresenter->present($this->context->cart),
>                 'errors' => empty($this->updateOperationError) ? '' : reset($this->updateOperationError),
> ```

Why does it need to be a string if no errors ? 

---

*[controllers/front/CartController.php, line 289 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPuw89mJgiZlugCLMe:-KyPuw89mJgiZlugCLMf:b-imhlgu) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L289)):*
> ```PHP
> 
>             $isAvailable = $this->shouldProductsAvailable();
>             if ($isAvailable !== true) {
> ```

Yoda condition ?

---

*[controllers/front/CartController.php, line 487 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPw9oKuBSfkIDg_T2g:-KyPw9oLL-nFbOJkw2VZ:b-p6ih36) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L487)):*
> ```PHP
>      * @return bool
>      */
>     private function shouldAvailabilityErrorBeRaised($product, $qty_to_check)
> ```

Please use camelCase notation for new variables / method parameters

---

*[controllers/front/CartController.php, line 504 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPrCHNEwgpGaEsIYi0:-KyPrCHOnUn3a6YhlB6k:be31qk) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L504)):*
> ```PHP
>      * @return bool|string
>      */
>     private function shouldProductsAvailable()
> ```

Wrong naming. Better "areProductsAvailable()"

---

*[controllers/front/CartController.php, line 508 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPwLeSWYktpqcc_Mlu:-KyPwLeT_Yhj9qoxJbnh:bw4sdkj) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L508)):*
> ```PHP
>         $product = $this->context->cart->checkQuantities(true);
> 
>         if ($product !== true && is_array($product)) {
> ```

Yoda condition

---

*[controllers/front/CartController.php, line 522 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPyfomN76LuQTfRPSS:-KyPyfon8VZ-LciMPOfX:brlusm7) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L522)):*
> ```PHP
>         }
> 
>         return true;
> ```

I suggest setting early returns to get rid of some indentation and annoying else :
```php
if (true === product || !is_array($product)) {
    return true;
}

if ($product['active']) {
    return $this->trans(/* qty alert */);
}

return $this->trans(/* availability alert */);
```

---

*[themes/classic/_dev/js/cart.js, line 29 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPxSqmyoBDI8NCy4xC:-KyPxSqmyoBDI8NCy4xD:b-m85kd9) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/themes/classic/_dev/js/cart.js#L29)):*
> ```JavaScript
>     });
>   });
>   
> ```

trailing white spaces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Review by LittleBigDev — DISMISSED** [review]
Thank you @kompilorb 
Please see my review above.

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 85 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPrW8EybfdUKw7Wcrs:-KyQswtJWNEYORuSMBlv:b-896fjx) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L85)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

I would change the order, and get rid of the `elseif` :
```php
if (Tools::getIsset('checkout')) {
    return Tools::redirect($this->context->link->getPageLink('order'));
}

if (true !== $isAvailable) { // Yoda condition
    $this->errors[] = $isAvailable;
}
```
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 289 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPuw89mJgiZlugCLMe:-KyQtWcRsJiQTWGxJIwj:bb6un1j) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L289)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Yoda condition ?
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 487 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPw9oKuBSfkIDg_T2g:-KyQtnZbreg13B8l9dLW:bb7f2mf) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L487)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Please use camelCase notation for new variables / method parameters
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 508 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPwLeSWYktpqcc_Mlu:-KyQu8CwX3JUwN1fMniv:b-7rl0kd) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L508)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Yoda condition
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 522 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPyfomN76LuQTfRPSS:-KyQuwKx-rl05zfEwTdm:bb7f2mf) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L522)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

I suggest setting early returns to get rid of some indentation and annoying else :
```php
if (true === product || !is_array($product)) {
    return true;
}

if ($product['active']) {
    return $this->trans(/* qty alert */);
}

return $this->trans(/* availability alert */);
```
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

File: controllers/front/CartController.php
Comment (Quetzacoalt91) [inline controllers/front/CartController.php]: You can fix an issue at the same time here. All variables to apply on a translated string must be associated to a placeholder.

```php
                        $this->trans(
                            'You must add %quantity% minimum quantity',
                            array('%quantity%' => $minimal_quantity),
                            [...]
```

**Review by Quetzacoalt91 — DISMISSED** [review]

File: controllers/front/CartController.php
Comment (kompilorb) [reply controllers/front/CartController.php]: Done :)

**Comment by kompilorb** [comment]
*[controllers/front/CartController.php, line 136 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPunTspzXxSgqYwM4e:-KypzDlajOxYLS1g1ar_:bs7wumb) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/controllers/front/CartController.php#L136)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Why does it need to be a string if no errors ? 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by kompilorb** [comment]
*[themes/classic/_dev/js/cart.js, line 29 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8478#-KyPxSqmyoBDI8NCy4xC:-KypzKH6AdS3nrNtIzg_:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/340245723eef9795d60a2271318c720ee1382f7b/themes/classic/_dev/js/cart.js#L29)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

trailing white spaces
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8478)*

**Comment by vincentbz** [comment]
Hi @toutantic and @LittleBigDev ,

Could it be possible to add the "waiting for QA feedback" before the merge, so that it can be tested ?

Thanks

**Comment by marionf** [comment]
Hello @kompilorb & @toutantic 
When the product is disabled, the displayed message is "This product (T-shirt délavé à manches courtes) is no longer available."
If I try to change the quantity, the message change for "This product is no longer available."
We want to keep the same error message: "This product (T-shirt délavé à manches courtes) is no longer available."

@LouiseBonnard, these strings aren't in Crowdin for 1.7: 
"This product (%product%) is no longer available."
"The item %product% in your cart is no longer available in this quantity. You cannot proceed with your order until the quantity is adjusted."

**Comment by kyripapa** [comment]
I seem to fix the comment on cart page however it still allows checkout. Has anyone else seen this?

**Pull request merged**
